nilearn/decoding/decoder.py

```diff
@@ -0,0 +1,840 @@
+        computed from masking.compute_background_mask or
+        masking.compute_epi_mask.
+
+    memory: instance of joblib.Memory or str
+        Used to cache the masking process.
+        By default, no caching is done. If a str is given, it is the
+        path to the caching directory.
+
+    memory_level: int, optional. Default: 0
+        Rough estimator of the amount of memory used by caching. Higher value
+        means more memory for caching.
+
+    n_jobs: int, optional. Default: 1.
+        The number of CPUs to use to do the computation. -1 means
+        'all CPUs'.
+
+    verbose: int, optional. Default: 0.
+        Verbosity level.
+
+    See Also
+    ------------
+    nilearn.decoding.DecoderRegressor: Regression strategies for Neuroimaging,
+    nilearn.decoding.Decoder: Classification strategies for Neuroimaging,
+    nilearn.decoding.SpaceNetClassifier: Graph-Net and TV-L1 priors/penalties
+    """
+
+    def __init__(self, estimator='svc', mask=None, cv=10, param_grid=None,
+                 screening_percentile=20, scoring=None, smoothing_fwhm=None,
+                 standardize=True, target_affine=None, target_shape=None,
+                 low_pass=None, high_pass=None, t_r=None,
+                 mask_strategy='background', is_classification=True,
+                 memory=None, memory_level=0, n_jobs=1, verbose=0):
+        self.estimator = estimator
+        self.mask = mask
+        self.cv = cv
+        self.param_grid = param_grid
+        self.screening_percentile = screening_percentile
+        self.scoring = scoring
+        self.is_classification = is_classification
+        self.smoothing_fwhm = smoothing_fwhm
+        self.standardize = standardize
+        self.target_affine = target_affine
+        self.target_shape = target_shape
+        self.mask_strategy = mask_strategy
+        self.low_pass = low_pass
+        self.high_pass = high_pass
+        self.t_r = t_r
+        self.memory = memory
+        self.memory_level = memory_level
+        self.n_jobs = n_jobs
+        self.verbose = verbose
+
+    def fit(self, X, y, groups=None):
+        """Fit the decoder (learner).
+
+        Parameters
+        ----------
+        X: list of Niimg-like objects
+            See http://nilearn.github.io/manipulating_images/input_output.html
+            Data on which model is to be fitted. If this is a list,
+            the affine is considered the same for all.
+
+        y: numpy.ndarray of shape=(n_samples) or list of length n_samples
+            The dependent variable (age, sex, IQ, yes/no, etc.).
+            Target variable to predict. Must have exactly as many elements as
+            3D images in niimg.
+
+        groups: None or bool, optional
+            Must be specify in multi-label classification settings.
+            Group labels for the samples used while splitting the dataset into
+            train/test set. This ‘groups’ parameter must always be specified to
```

Comment:
what do you mean by this? it must not always be specified if it is optional

Reply:
Also, how is setting `groups` to `None` different from setting `groups` to `False`?

Reply:
Ok this parameter cannot be set to `bool` type, see: https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.StratifiedKFold.html#sklearn.model_selection.StratifiedKFold.split

Reply:
Based on that link, groups is never used by Scikit-learn. I think we don't need to maintain that compatibility (I may be wrong) so we can simply remove this as a user facing parameter.

Reply:
And in fact it can be set to any type (it says type object and everything in Python is n object).
It's never used.

Reply:
Note that this is a crucial thing for Decoder, since it is where you specify session or subject information, that is mandatory for having good splits.

Reply:
The scikit-learn documentation says it's not used. Is this purely for adding more descriptive information?
I thought the cv object (`LeaveOneGroupOut()` here) handled the groups.

Reply:
@tbng mentions the code is running fine without the `groups` parameter.

Reply:
The same link https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.StratifiedKFold.html#sklearn.model_selection.StratifiedKFold.split specifies that

> Note that providing y is sufficient to generate the splits

so maybe the API has changed since then and specifying `groups` is not neccessary anymore.

Reply:
Maybe, but not specifying groups is almost surely very bad, because you end up having non-independent test sets, hence biased accuracies !

Reply:
OKay, then rewriting this docstring entry will resolve this.

Reply:
OK there seems to be ambiguity in the docstring. By default (if the `cv` parameter is `int`) the Decoder object will initialized cross-validation object as `sklearn.model_selection.StratifiedKFold`, which does not need to specify the `groups` when using `split` attribute. 

However if we input some cv generators like `sklearn.model_selection.LeaveOneGroupOut` the groups parameter must be specified as group labels in order to split the data into train as test set, for the reason @bthirion mentioned.

I will edit the docstring with more details.